qcodes/utils/logger.py

```diff
@@ -0,0 +1,439 @@
+                  handler: Optional[
+                           Union[logging.Handler,
+                                 Sequence[logging.Handler]]]=None):
+    """
+    Context manager to temporarily change the level of handlers.
+    Example:
+        >>> with logger.handler_level(level=logging.DEBUG, handler=[h1, h1]):
+        >>>     root_logger.debug('this is now visible)
+
+    Args:
+        level: level to set the handlers to
+        handler: single or sequence of handlers which to change
+    """
+    if isinstance(handler, logging.Handler):
+        handler = (handler,)
+    original_levels = [h.level for h in handler]
+    for h in handler:
+        h.setLevel(level)
+    yield
+    for h, original_level in zip(handler, original_levels):
+        h.setLevel(original_level)
+
+
+@contextmanager
+def console_level(level: LevelType):
+    """
+    Context manager to temporarily change the level of the qcodes console
+    handler.
+    Example:
+        >>> with logger.console_level(level=logging.DEBUG):
+        >>>     root_logger.debug('this is now visible)
+
+    Args:
+        level: level to set the console handler to
+    """
+    global console_handler
+    with handler_level(level, handler=console_handler):
+        yield
+
+
+@contextmanager
+def filter_instrument(instrument: Union[InstrumentBase,
+                                        Sequence[InstrumentBase]],
+                      handler: Optional[
+                          Union[logging.Handler,
+                                Sequence[logging.Handler]]]=None,
+                      level: Optional[LevelType]=None):
+    """
+    Context manager that adds a filter that only enables the log messages of
+    the supplied instruments to pass.
+    Example:
+        >>> h1, h2 = logger.get_console_handler(), logger.get_file_handler()
+        >>> with logger.filter_instruments((qdac, dmm2), handler=[h1, h1]):
+        >>>     qdac.ch01(1)  # logged
+        >>>     v = dmm2.v()  # not logged
+
+    Args:
+        level: level to set the handlers to
+        handler: single or sequence of handlers which to change
+    """
+    if handler is None:
+        global console_handler
+        handler = (console_handler,)
+    if isinstance(handler, logging.Handler):
+        handler = (handler,)
+
+    instrument_filter = InstrumentFilter(instrument)
+    for h in handler:
+        h.addFilter(instrument_filter)
+    if level is not None:
+        with handler_level(level, handler):
+            yield
+    else:
+        yield
+    for h in handler:
+        h.removeFilter(instrument_filter)
+
+
+@contextmanager
+def capture_dataframe(level: LevelType=logging.DEBUG,
+                      logger: logging.Logger=None):
+    """
+    Context manager to capture the logs in a `pandas.DataFrame`
+
+    Example:
+        >>> with logger.capture_dataframe() as (handler, cb):
+        >>>     qdac.ch01(1)  # some commands
+        >>>     data_frame = cb()
```

Comment:
if one calls `cb` multiple times, will it return the complete log at the moment of the call, or the log captured between previous call and this call?